File: Workshop 3/src/BlackJack/model/Player.java
/*
 * Player.java
 * Date: 26 Oktober, 2014
 * Author: Tobias Ohlsson (edited by Robert Willem Hallink)
 */
package BlackJack.model;
import java.util.List;
import java.util.LinkedList;

/*
 * Mostly left untouched.
 * Code regarding hidden cards is adjusted (see Card.java).
 */
public class Player {
    
    private final List<Card> m_hand;
    protected final int g_maxScore = 21;
    
    public Player() {
  
        m_hand = new LinkedList<>();
        
    }
    
    public void DealCard(Card a_addToHand)      { m_hand.add(a_addToHand); }
    public Iterable<Card> GetHand()             { return m_hand; }
    public void ClearHand()                     { m_hand.clear(); }
    
    public void ShowHand() {
        
        m_hand.stream().forEach((c) -> {
            c.Show(true);
        });
        
    }
    
    public int CalcScore() {
        
        // the number of scores is dependant on the number of scorable values
        // as it seems there is no way to do this check at compile time in java ?!
        // cardScores[13] = {...};
        int cardScores[] = { 2, 3, 4, 5, 6, 7, 8, 9, 10, 10 ,10 ,10, 11 };
        assert (cardScores.length == Card.Value.Count.ordinal()) : "Card Scores array size does not match number of card values";
        
        int score = 0;
        
        for (Card c : GetHand())
            if (!c.IsHidden())
                score += cardScores[c.GetValue().ordinal()];

        if (score > g_maxScore)
            for (Card c : GetHand())
                if (c.GetValue() == Card.Value.Ace && score > g_maxScore)
                    score -= 10;
        
        return score;
        
    }
    
}

File: Workshop 3/Files/Peer Review/Materials - Eddy Proca/model/Player.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace BlackJack.model
{
    class Player
    {
        private List<Card> m_hand = new List<Card>();
        private int softHand;
        public void DealCard(Card a_card)
        {
            m_hand.Add(a_card);
        }
        public IEnumerable<Card> GetHand()
        {
            return m_hand.Cast<Card>();
        }
        public void ClearHand()
        {
            m_hand.Clear();
        }
        public void ShowHand()
        {
            foreach (Card c in GetHand())
            {
                c.Show(true);
            }
        }
        public int CalcScore()
        {
            softHand = 0;
            int[] cardScores = new int[(int)model.Card.Value.Count]
                {2, 3, 4, 5, 6, 7, 8, 9, 10, 10 ,10 ,10, 11};
            int score = 0;

            foreach(Card c in GetHand()) {
                if (c.GetValue() != Card.Value.Hidden)
                {
                    score += cardScores[(int)c.GetValue()];
                }
                if (c.GetValue() == Card.Value.Ace)
                {
                    softHand++;
                }
            }
            if (score > 21)
            {
                foreach (Card c in GetHand())
                {
                    if (c.GetValue() == Card.Value.Ace && score > 21)
                    {
                        score -= 10;
                        softHand--;
                    }
                }
            }
            return score;
        }
        public bool SoftHand() 
        {
            if (softHand == 0)
            {
                return false;
            }
            return true;
        }
    }
}
